### ISSUE TITLE
Bert Model prediction giving same output

### ISSUE THREAD

**Issue by atirpetkar** [as-ideas/headliner#18]
Hi @cschaefer26 and @datitran ,

Thank you so much for the headliner library, I have been playing around it for a while and so far really enjoyed it.

I was working on a use case for summarization and use headliner's bert model, I followed the following readme code (with few tweaks for SummarizerTransformer parameters) for headliner with my use case train_data:

```python
from headliner.preprocessing import Preprocessor

train_data = [('Some inputs.', 'Some outputs.')] * 10

# use BERT-specific start and end token
preprocessor = Preprocessor(start_token='[CLS]',
                            end_token='[SEP]',
                            lower_case=True)
train_prep = [preprocessor(t) for t in train_data]
targets_prep = [t[1] for t in train_prep]


from tensorflow_datasets.core.features.text import SubwordTextEncoder
from transformers import BertTokenizer
from headliner.model import SummarizerBert

# Use a pre-trained BERT embedding and BERT tokenizer for the encoder 
tokenizer_input = BertTokenizer.from_pretrained('bert-base-uncased')
tokenizer_target = SubwordTextEncoder.build_from_corpus(
    targets_prep, target_vocab_size=2**13,  reserved_tokens=[preprocessor.start_token, preprocessor.end_token])

vectorizer = Vectorizer(tokenizer_input, tokenizer_target)
summarizer = SummarizerBert(num_heads=4,
                            feed_forward_dim=512,
                            num_layers_encoder=3,
                            num_layers_decoder=3,
                            bert_embedding_encoder='bert-base-uncased',
                            embedding_encoder_trainable=False,
                            embedding_size_encoder=768,
                            embedding_size_decoder=64,
                            dropout_rate=0.1,
                            max_prediction_len=400)
)
summarizer.init_model(preprocessor, vectorizer)

trainer = Trainer(batch_size=2)
trainer.train(summarizer, train_data, num_epochs=200)
```
I train the model for 200 epochs and i see in logs that the loss keeps on reducing (starting from around 4 to 0.69), so seems like training happens just fine.

After that when i try to do prediction on the saved model using following code, it gives me the same prediction always for any test_sentence :

```Python
from headliner.model.summarizer_bert import SummarizerBert

summarizer = SummarizerBert.load('/path/to/headliner_bert_model')
summarizer.predict(test_sentence)
```

Please if anyone can advise if am missing anything? with respect to prediction part for bert model?

**Comment by cschaefer26** [comment]
Hi, this shouldn't happen. Did you see in the logs that the current model is saved correctly? I would suggest to produce a train/val split and do trainer.train(train_data, val_data=val_data), the output should be more verbose during training.

**Comment by atirpetkar** [comment]
Hi @cschaefer26 ,

Thanks for your reply.
As seen in the attached screenshot which shows a snippet of the end of my headliner-bert model training log, which shows i save and then also load the model from same path. So i think so the model is getting saved correctly.
 
<img width="1503" alt="Screen Shot 2019-12-20 at 12 36 56 PM" src="https://user-images.githubusercontent.com/13209770/71278429-96dc3f00-2325-11ea-9ddf-c13545ccd50f.png">

I will try checking using val_data as well and see if the situation still persists.

**Comment by cschaefer26** [comment]
Sounds good, whats your training data?  Hopefully not the one in the script above ;-)

**Comment by atirpetkar** [comment]
Hi @cschaefer26 ,
haha yes the training data is not the one as provided in the readme, but it is regularity obligation data, for which we have both actual text (X) and its summarized version (target). So i was trying to use headliner for that. I successfully used SummarizerTransformer, but only got around 0.20 rouge-1 f score :(, so i wanted to try SummarizerBert as well.

Also another issue, other than the one I discussed above is the actual text being more than **512 tokens** sometimes in train_data and hence I am unable to use bert model for those instances. 
Any idea in your experience how to handle for them as well?

**Comment by cschaefer26** [comment]
I see - for good results I would suggest to use >4 layers and embedding size 768 and 8 heads. If you want to use the Bert model I recommend to make the Bert embedding trainable too and use Adam with a lower learning rate (i.e. summarizer.optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4)). With Bert unfortunately there is no possibility to process >512 tokens, but you can set vectorizer.max_input_len = 512 and then the encoded sequences will be cropped at 512. Good luck.

**Comment by cschaefer26** [comment]
Hi, any news on the problem? I recently updated the bert model with some big improvements, might be worth a try.
